File: multilibrary/src/main/java/com/zhongjh/albumcamerarecorder/album/base/BaseRecyclerViewCursorAdapter.java
/*
 * Copyright 2017 Zhihu Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.zhongjh.albumcamerarecorder.album.base;

import android.database.Cursor;
import android.provider.MediaStore;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

/**
 * 父类
 *
 * @author zhongjh
 * @param <VH>
 */
public abstract class BaseRecyclerViewCursorAdapter<VH extends RecyclerView.ViewHolder> extends
        RecyclerView.Adapter<VH> {

    private Cursor mCursor;
    private int mRowIdColumn;

    protected BaseRecyclerViewCursorAdapter(Cursor c) {
        setHasStableIds(true);
        swapCursor(c);
    }

    /**
     * 绑定数据源
     *
     * @param holder 控件
     * @param cursor 加载完成的游标
     */
    protected abstract void onBindViewHolder(VH holder, Cursor cursor);

    @Override
    public void onBindViewHolder(@NonNull VH holder, int position) {
        if (!isDataValid(mCursor)) {
            throw new IllegalStateException("Cannot bind view holder when cursor is in invalid state.");
        }
        if (!mCursor.moveToPosition(position)) {
            throw new IllegalStateException("Could not move cursor to position " + position
                    + " when trying to bind view holder");
        }

        onBindViewHolder(holder, mCursor);
    }

    @Override
    public int getItemViewType(int position) {
        if (!mCursor.moveToPosition(position)) {
            throw new IllegalStateException("Could not move cursor to position " + position
                    + " when trying to get item view type.");
        }
        return getItemViewType(position, mCursor);
    }

    /**
     * 返回类型
     * @param position 索引
     * @param cursor   游标
     * @return 返回类型
     */
    protected abstract int getItemViewType(int position, Cursor cursor);

    @Override
    public int getItemCount() {
        if (isDataValid(mCursor)) {
            return mCursor.getCount();
        } else {
            return 0;
        }
    }

    @Override
    public long getItemId(int position) {
        if (!isDataValid(mCursor)) {
            throw new IllegalStateException("Cannot lookup item id when cursor is in invalid state.");
        }
        if (!mCursor.moveToPosition(position)) {
            throw new IllegalStateException("Could not move cursor to position " + position
                    + " when trying to get an item id");
        }

        return mCursor.getLong(mRowIdColumn);
    }

    /**
     * 填充数据源
     *
     * @param newCursor 数据源
     */
    public void swapCursor(Cursor newCursor) {
        if (newCursor == mCursor) {
            return;
        }

        if (newCursor != null) {
            mCursor = newCursor;
            mRowIdColumn = mCursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns._ID);
            // notify the observers about the new cursor
            notifyDataSetChanged();
        } else {
            notifyItemRangeRemoved(0, getItemCount());
            mCursor = null;
            mRowIdColumn = -1;
        }
    }

    protected Cursor getCursor() {
        return mCursor;
    }

    private boolean isDataValid(Cursor cursor) {
        return cursor != null && !cursor.isClosed();
    }
}

File: multilibrary/src/main/java/com/zhongjh/albumcamerarecorder/album/base/BaseRecyclerViewCursorAdapter.kt
package com.zhongjh.albumcamerarecorder.album.base

import android.database.Cursor
import android.provider.MediaStore
import androidx.recyclerview.widget.RecyclerView

/**
 * 父类
 *
 * @author zhongjh
 * @param <VH>
</VH> */
abstract class BaseRecyclerViewCursorAdapter<VH : RecyclerView.ViewHolder?>  :
    RecyclerView.Adapter<VH>() {
    protected var cursor: Cursor? = null
        private set
    private var mRowIdColumn = 0

    /**
     * 绑定数据源
     *
     * @param holder 控件
     * @param cursor 加载完成的游标
     */
    protected abstract fun onBindViewHolder(holder: VH, cursor: Cursor?)
    override fun onBindViewHolder(holder: VH, position: Int) {
        check(isDataValid(cursor)) { "Cannot bind view holder when cursor is in invalid state." }
        check(cursor!!.moveToPosition(position)) {
            ("Could not move cursor to position " + position
                    + " when trying to bind view holder")
        }
        onBindViewHolder(holder, cursor)
    }

    override fun getItemViewType(position: Int): Int {
        check(cursor!!.moveToPosition(position)) {
            ("Could not move cursor to position " + position
                    + " when trying to get item view type.")
        }
        return getItemViewType(position, cursor)
    }

    /**
     * 返回类型
     * @param position 索引
     * @param cursor   游标
     * @return 返回类型
     */
    protected abstract fun getItemViewType(position: Int, cursor: Cursor?): Int
    override fun getItemCount(): Int {
        return if (isDataValid(cursor)) {
            cursor!!.count
        } else {
            0
        }
    }

    override fun getItemId(position: Int): Long {
        check(isDataValid(cursor)) { "Cannot lookup item id when cursor is in invalid state." }
        check(cursor!!.moveToPosition(position)) {
            ("Could not move cursor to position " + position
                    + " when trying to get an item id")
        }
        return cursor!!.getLong(mRowIdColumn)
    }

    /**
     * 填充数据源
     *
     * @param newCursor 数据源
     */
    fun swapCursor(newCursor: Cursor?) {
        if (newCursor === cursor) {
            return
        }
        if (newCursor != null) {
            cursor = newCursor
            mRowIdColumn = cursor!!.getColumnIndexOrThrow(MediaStore.Files.FileColumns._ID)
            notifyDataSetChanged()
        } else {
            notifyItemRangeRemoved(0, itemCount)
            cursor = null
            mRowIdColumn = -1
        }
    }

    private fun isDataValid(cursor: Cursor?): Boolean {
        return cursor != null && !cursor.isClosed
    }

    init {
    }
}